ori/volume/server/volume_create.go

```diff
@@ -16,261 +16,143 @@ package server
-	if err != nil {
-		return nil, fmt.Errorf("failed to resolve image ref in registry: %w", err)
-	}
-
-	layers, err := img.Layers(ctx)
-	if err != nil {
-		return nil, fmt.Errorf("failed to get layers for image: %w", err)
-	}
+func (s *Server) getAggregatedImage(ctx context.Context, log logr.Logger, volume *ori.Volume) (*aggregatedImage, error) {
+	log.V(2).Info("Starting volume validation")
+	defer log.V(2).Info("Finished volume validation")
 
-	var rootFSLayer image.Layer
-	for _, l := range layers {
-		if l.Descriptor().MediaType == onmetalimage.RootFSLayerMediaType {
-			rootFSLayer = l
-			break
-		}
-	}
-	if rootFSLayer == nil {
-		return nil, fmt.Errorf("failed to get rootFS layer")
+	if volume == nil {
+		return nil, fmt.Errorf("volume is nil")
 	}
 
-	size := resource.NewQuantity(rootFSLayer.Descriptor().Size, resource.BinarySI)
-	if size == nil {
-		return nil, fmt.Errorf("failed to get size of rootFS layer")
+	class, found := s.volumeClasses.Get(volume.Spec.Class)
+	if !found {
+		return nil, fmt.Errorf("volume class '%s' not supported", volume.Spec.Class)
 	}
+	log.V(2).Info("Validated class")
 
-	return size, nil
-}
+	calculatedLimits := limits.Calculate(class.Capabilities.Iops, class.Capabilities.Tps, s.burstFactor, s.burstDurationInSeconds)
 
-func (s *Server) getVolumeConfig(ctx context.Context, volume *ori.Volume) (*AggregateCephVolume, error) {
-	volumeClass := &storagev1alpha1.VolumeClass{}
-	if err := s.client.Get(ctx, types.NamespacedName{Name: volume.Spec.Class}, volumeClass); err != nil {
-		return nil, fmt.Errorf("unable to get VolumeClass: %w", err)
-	}
-
-	pvc := &corev1.PersistentVolumeClaim{
-		TypeMeta: metav1.TypeMeta{
-			Kind:       "PersistentVolumeClaim",
-			APIVersion: "v1",
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      s.idGen.Generate(),
-			Namespace: s.namespace,
-		},
-		Spec: corev1.PersistentVolumeClaimSpec{
-			AccessModes: []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
-			Resources: corev1.ResourceRequirements{
-				Requests: corev1.ResourceList{corev1.ResourceStorage: *resource.NewQuantity(int64(volume.Spec.Resources.StorageBytes), resource.DecimalSI)},
-			},
-			VolumeMode:       func(m corev1.PersistentVolumeMode) *corev1.PersistentVolumeMode { return &m }(corev1.PersistentVolumeBlock),
-			StorageClassName: &s.rookPoolStorageClass,
-		},
+	labels := volume.Metadata.Labels
+	if labels == nil {
+		labels = map[string]string{}
 	}
 
-	var imagePvc *corev1.PersistentVolumeClaim
-	if volume.Spec.Image != "" {
-		imageName := GetSanitizedImageName(volume.Spec.Image)
-		apiutils.SetImageAnnotation(pvc, volume.Spec.Image)
+	labels[VolumeClassLabel] = volume.Spec.Class
 
-		pvc.Spec.DataSourceRef = &corev1.TypedObjectReference{
-			APIGroup: pointer.String("snapshot.storage.k8s.io"),
-			Kind:     "VolumeSnapshot",
-			Name:     imageName,
+	var snapshotRef *string
+	var snapshot *api.Snapshot
+	if image := volume.Spec.Image; image != "" {
+		img, err := s.registry.Resolve(ctx, image)
```

Comment:
The image shouldn't be resolved at creation time. If an image e.g. becomes available later, the volume should stay in state = Pending and then be populated once available. Your process of creating the snapshot / using it will have to be adjusted to fit this case.